pkg/quotaplugins/quota-forest/quota-manager/quota/quotamanager.go

```diff
@@ -269,26 +269,38 @@ func (m *Manager) Allocate(treeName string, consumerID string) (response *core.A
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	agent := m.agents[treeName]
-	if agent == nil {
-		return nil, fmt.Errorf("invalid tree name %s", treeName)
+	agent, consumer, err := m.preAllocate(treeName, consumerID)
+	if err == nil && agent.controller.IsConsumerAllocated(consumerID) {
+		err = fmt.Errorf("consumer %s already allocated on tree %s", consumerID, treeName)
 	}
-	consumerInfo := m.consumerInfos[consumerID]
-	if consumerInfo == nil {
-		return nil, fmt.Errorf("consumer %s does not exist, create and add first", consumerID)
+	if err != nil {
+		return nil, err
 	}
-	if agent.controller.IsConsumerAllocated(consumerID) {
-		return nil, fmt.Errorf("consumer %s already allocated on tree %s", consumerID, treeName)
+	if m.mode == Normal {
+		response = agent.controller.Allocate(consumer)
+	} else {
+		response = agent.controller.ForceAllocate(consumer, consumer.GetGroupID())
 	}
-	resourceNames := agent.cache.GetResourceNames()
-	consumer, err := consumerInfo.CreateTreeConsumer(treeName, resourceNames)
-	if err != nil {
-		return nil, fmt.Errorf("failure creating consumer %s on tree %s", consumerID, treeName)
+	if !response.IsAllocated() {
+		return nil, fmt.Errorf(response.GetMessage())
 	}
+	return response, err
+}
 
-	klog.V(4).Infoln(consumer)
+// TryAllocate : try allocating a consumer on a tree
+func (m *Manager) TryAllocate(treeName string, consumerID string) (response *core.AllocationResponse, err error) {
```

Comment:
@atantawi would be possible to cover the new methods with unit tests? There is a `pkg/quotaplugins/quota-forest/quota-manager/quota/quotamanager_test.go`. We should try to test in a concurent mannaer.

Reply:
added unit test for TreeController

Reply:
That test does not cover the concurrency issues, I think we should still augment the tests to cover the new functionality.

Reply:
Added concurrent tests in `quotamanagerundo_test.go`

Reply:
Ok. The explanation, makes sense.